lib/linked_list.rb

```diff
@@ -19,99 +19,271 @@ def initialize
+        @head = current.next
+      else
+        previous.next = current.next
+      end
+      
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
     def reverse
-      raise NotImplementedError
+      return if @head.nil?
+      current = @head
+      previous = nil
+
+      until current.next.nil?
+        temp = current.next
+        current.next = previous
+        previous = current
+        current = temp
+      end
+
+      current.next = previous
+      @head = current
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
     def find_middle_value
-      raise NotImplementedError
+      length = self.length
+
+      return nil if @head.nil?
+
+      value = (length / 2)
+
+      return self.get_at_index(value)
+      #OR...... this way is faster
+      # fast = @root
+      # slow = @root
+      # until fast.nil?
+      #   fast = fast.next
+      #   fast = fast.next unless fast.nil?
+      #   slow = slow.next
+      # end
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     def find_nth_from_end(n)
-      raise NotImplementedError
+      current = @head
+      n_ahead = @head
+
+      return nil if current.nil?
+
+      n.times do
+        return nil if n_ahead.next.nil?
+        n_ahead = n_ahead.next
+      end
+
+      until n_ahead.next.nil?
+        return current.data if current.next.nil?
+        current = current.next
+        n_ahead = n_ahead.next
+      end
+      return current.data
+      # return get_at_index(length() - 1 - n)
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     def has_cycle
-      raise NotImplementedError
-    end
+      slow = @head
+      fast = @head
+      fast = fast.next if fast
 
+      until fast.nil? || slow.nil?
+        if fast == slow
+          return true
+        end
+        fast = fast.next
+        return true if fast == slow
+        fast = fast.next unless fast.nil?
+        slow = slow.next
+      # until fast.nil?
+      #   slow = slow.next
+      #   fast = fast.next
+      #   fast = fast.next unless fast.nil?
+      #   if slow == fast
+      #     return true
+      #   end
+      end
+      return false
+    end
 
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
     def get_first
-      raise NotImplementedError
+      if @head.nil?
+        return nil
+      else
+        return @head.data
+      end
     end
 
     # method that inserts a given value as a new last node in the linked list
     def add_last(value)
-      raise NotImplementedError
+      current = @head
+
+      if current.nil?
+        return Node.new(value)

```

Comment:
You're not changing `@head` here.

```suggestion
        @head = Node.new(value)
```